docs/specifications/tests/DNSSEC-TP/dnssec02.md

```diff
@@ -3,125 +3,202 @@
+    7.  Name server IP address, DNSKEY record key tag and DNSKEY algorithm code
+        ("Algo Not Supported By ZM").
+    8. Name server IP and key tag from RRSIG record ("RRSIG not valid by DNSKEY").
+
+2.  If the *Test Type* is "[undelegated]" do:
+    1. If *Undelegated DS* is empty then do terminate this test case.
+    2. Else add *Undelegated DS* as DS records to the *DS Record* set.
+
+3.  If *Test Type* is "normal", then:
+    1. Create a DS query with the DO flag set for the name of the *Child Zone*
+       ("DS Query").
+    2. Retrieve all name server IP addresses for the parent zone of
+       *Child Zone* using [Method1] ("Parent NS IP").
+    3. For each parent name server in *Parent NS IP* do:
+       1. Send *DS Query* to the name server IP.
+       2. If at least one of the following criteria is met, then go to next
+          parent name server:
+          1. There is no DNS response.
+          2. The RCODE of response is not "NoError" ([IANA RCODE List]).
+          3. The AA flag is not set in the response.
+          4. There is no DS record with matching owner name in the answer
+             section.
+       3. Retrieve the DS records from the response and add them to the
+          *DS Record* set.
+    4. If the *DS Record* set is empty exit this test case.
+4.  Create a DNSKEY query for the *Child Zone* with the DO flag set
+    ("DNSKEY Query").
+
+5.  Obtain the set of child name server IP addresses using [Method4] and
+    [Method5] ("Child NS IP").
+
+6.  For each child name server in *Child NS IP* do:
+    1. Send *DNSKEY Query* over UDP to the name server IP and collect the response.
+    2. If at least one of the following criteria is met, then go to next
+          child name server:
+       1. There is no DNS response.
+       2. The RCODE of the response is not "NoError" ([IANA RCODE List]).
+       3. The AA flag is not set in the response.
+       4. There is no DNSKEY record with owner name matching the query in the
+          answer section.
+    3. Extract the DNSKEY RRset ("DNSKEY RRs").
+    4. Extract the RRSIG records covering the DNSKEY RRset, possibly
+       none ("DNSKEY RRSIG").
+    5. For each DS in *DS Records*, do:
+       1. Find the equivalent DNSKEY in *DNSKEY RRs* by key ID (key tag).
+       2. If matching DNSKEY is not found add DS key tag and name server IP to
+          the *No DNSKEY for DS* set.
+       3. If the DS values (algorithm and digest) do not match the DNSKEY record
+          then add DS key tag and name server IP to the *No match DS DNSKEY* set.
+       4. If bit 7 of the DNSKEY flags field is unset (value 0), then add DS key
+          tag and name server IP to the *DNSKEY not for zone signing* set.
+       5. If bit 15 of the DNSKEY flags field is unset (value 0), then add the
+          DNSKEY record key tag and name server IP to the *DNSKEY not SEP*
+          set.
+       6. Find the equivalent RRSIG in *DNSKEY RRSIG* by key ID (key tag).
```

Comment:
Could there be more than one RRSIG with the same key tag in *DNSKEY RRSIG*?

Reply:
Yes. If there are more than one DNSKEY record with the same key tag (see answer above) and more than one of them is used to create an RRSIG record that will happen.

Updated to handle that case.